File: backend/app/domains/chat/schemas/trust_metadata.py
"""Versioned answer trust metadata contract (F307).

This module defines stable, security-filtered DTOs for answer trust metadata.
The schema_version field allows frontend consumers to handle contract evolution
without breaking. Raw prompts, chain-of-thought, ACL snapshots, and internal
UUIDs are excluded.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


class CitationTrustRecord(BaseModel):
    """Trust-enriched citation record.

    Omits source_acl_snapshot (internal ACL details) and keeps only
    user-facing provenance and freshness fields.

    F311 additions:
      freshness_state      — normalized display state for the trust panel
      doc_last_updated_at  — when the document was last modified
      doc_review_owner_id  — who is responsible for reviewing this document
      doc_unreviewed_warning — True when the source is pending review
      doc_deprecated_warning — True when the source is deprecated/archived/superseded
    """

    document_id: str
    chunk_id: str
    filename: str | None = None
    page_number: int | None = None
    score: float | None = None
    similarity_score: float | None = None
    rerank_score: float | None = None
    original_rank: int | None = None
    final_rank: int | None = None
    text_snippet: str | None = None
    source_provider: str | None = None
    source_provider_label: str | None = None
    source_title: str | None = None
    source_key: str | None = None
    source_section: str | None = None
    source_deep_link: str | None = None
    source_last_synced_at: datetime | None = None
    source_trust_status: (
        Literal["trusted", "stale", "revoked", "deleted", "unknown", "uploaded"] | None
    ) = None
    conflict_status: Literal["preferred", "conflicting", "neutral"] | None = None
    doc_trust_status: str | None = None
    doc_quality_state: str | None = None
    doc_review_status: str | None = None
    doc_version_label: str | None = None
    doc_review_due_date: date | None = None
    doc_expiry_date: date | None = None
    doc_stale_warning: bool = False
    doc_expired_warning: bool = False
    doc_is_excluded_status: bool = False
    doc_draft_warning: bool = False
    is_table_chunk: bool = False
    table_caption: str | None = None
    table_row_count: int | None = None
    table_col_count: int | None = None
    table_headers: list[str] = Field(default_factory=list)
    doc_ocr_quality_status: str | None = None
    doc_ocr_low_confidence_warning: bool = False
    # F311 — normalized freshness state + provenance display fields
    freshness_state: (
        Literal["current", "stale", "expired", "deprecated", "draft", "unreviewed", "unknown"]
        | None
    ) = None
    doc_last_updated_at: datetime | None = None
    doc_review_owner_id: str | None = None
    doc_unreviewed_warning: bool = False
    doc_deprecated_warning: bool = False
    doc_draft_warning: bool = False
    # F315 — evidence quality: table extraction and document processing signals
    table_extraction_confidence: float | None = None
    table_low_confidence_warning: bool = False
    doc_extraction_quality: str | None = None
    doc_extraction_warning: bool = False
    doc_processing_warning: bool = False


class ConfidenceReasonRecord(BaseModel):
    """Single explainable signal that contributed to the confidence score (F310)."""

    code: str
    label: str
    impact: Literal["positive", "negative", "neutral"]
    magnitude: float = Field(ge=0.0, le=1.0)


class ConfidenceTrustRecord(BaseModel):
    """Confidence score breakdown for the answer."""

    score: float = Field(ge=0.0, le=1.0)
    category: Literal["low", "medium", "high"]
    trust_level: Literal["high", "medium", "low", "warning", "not_found"] = "low"
    citation_support_score: float = Field(ge=0.0, le=1.0)
    citation_validation_score: float = Field(ge=0.0, le=1.0)
    citation_coverage_score: float = Field(ge=0.0, le=1.0)
    retrieval_agreement_score: float = Field(ge=0.0, le=1.0)
    top_similarity: float = Field(ge=0.0, le=1.0)
    average_similarity: float = Field(ge=0.0, le=1.0)
    top_rerank_score: float = Field(ge=0.0, le=1.0)
    raw_score: float = Field(ge=0.0, le=1.0)
    citation_validation_multiplier: float = Field(ge=0.0, le=1.0)
    not_found_penalty_multiplier: float = Field(ge=0.0, le=1.0)
    freshness_multiplier: float = Field(ge=0.0, le=1.0, default=1.0)
    ocr_quality_multiplier: float = Field(ge=0.0, le=1.0, default=1.0)
    conflict_multiplier: float = Field(ge=0.0, le=1.0, default=1.0)
    table_quality_multiplier: float = Field(ge=0.0, le=1.0, default=1.0)
    extraction_quality_multiplier: float = Field(ge=0.0, le=1.0, default=1.0)
    graph_evidence_boost: float = Field(ge=0.0, le=1.0, default=0.0)
    verification_support_score: float | None = None
    not_found_signal: bool
    no_context: bool
    reasons: list[ConfidenceReasonRecord] = Field(default_factory=list)


class RetrievalDiagnosticsRecord(BaseModel):
    """Safe subset of pipeline retrieval diagnostics."""

    retrieval_candidate_count: int = 0
    retrieval_count: int = 0
    selected_count: int = 0
    top_k: int = 0
    search_mode: str | None = None
    source_scope_mode: str | None = None
    source_scope_label: str | None = None
    retrieval_profile_name: str | None = None
    retrieval_profile_scope: str | None = None
    retrieval_profile_source: str | None = None
    retrieval_filters: list[str] = Field(default_factory=list)
    rerank_applied: bool = False
    rerank_provider: str | None = None
    rerank_model: str | None = None
    rerank_score_min: float | None = None
    rerank_score_max: float | None = None
    rerank_fallback_used: bool = False
    rerank_fallback_reason: str | None = None
    request_id: str | None = None
    trace_request_id: str | None = None
    hybrid_retrieval_enabled: bool = False
    hybrid_vector_hit_count: int = 0
    hybrid_keyword_hit_count: int = 0
    query_rewriting_applied: bool = False
    query_decomposed: bool = False
    sub_query_count: int = 0
    parent_context_expanded_count: int = 0
    graph_context_used: bool = False
    graph_context_unavailable: bool = False
    graph_chunk_count: int = 0
    freshness_excluded_count: int = 0
    freshness_boosted_count: int = 0


class QueryInterpretationRecord(BaseModel):
    """Safe query interpretation surfaced in the trust panel."""

    intent: Literal[
        "lookup",
        "summary",
        "comparison",
        "policy",
        "troubleshooting",
        "compliance",
        "connector_search",
        "graph_entity_search",
    ]
    intent_label: str
    complexity: Literal["simple", "complex", "multi_part"]
    retrieval_strategy: Literal["original", "rewrite", "decompose"]
    rewrite_preview_enabled: bool = True
    rewritten_query_preview: str | None = None
    sub_queries: list[str] = Field(default_factory=list)


class GroundedVerificationRecord(BaseModel):
    """Results of the post-generation grounded answer verifier (F296/F338)."""

    applied: bool = False
    verdict: str | None = None
    score: float | None = None
    aggregate_support_score: float = Field(ge=0.0, le=1.0, default=0.0)
    claim_count: int = 0
    supported_count: int = 0
    partially_supported_count: int = 0
    unsupported_count: int = 0
    unverifiable_count: int = 0
    conflicting_count: int = 0
    not_enough_evidence_count: int = 0
    removed_count: int = 0
    reason_codes: list[str] = Field(default_factory=list)
    claims: list[ClaimSupportRecord] = Field(default_factory=list)
    mode: str | None = None
    threshold: float | None = None


class ClaimSupportRecord(BaseModel):
    """Claim-level support details with citation mapping."""

    claim_index: int = Field(ge=1)
    claim_text: str
    support_status: Literal[
        "supported",
        "partially_supported",
        "unsupported",
        "unverifiable",
        "conflicting",
        "not_enough_evidence",
    ]
    support_score: float = Field(ge=0.0, le=1.0)
    evidence_match_score: float = Field(ge=0.0, le=1.0)
    source_quality_score: float = Field(ge=0.0, le=1.0)
    rerank_score: float = Field(ge=0.0, le=1.0)
    chunk_coverage_score: float = Field(ge=0.0, le=1.0)
    citation_indices: list[int] = Field(default_factory=list)


class ModelMetadataRecord(BaseModel):
    """LLM and prompt template metadata.

    Excludes the internal prompt_template_version_id UUID.
    """

    llm_model: str | None = None
    llm_provider: str | None = None
    embedding_model: str | None = None
    fallback_used: bool = False
    fallback_from: str | None = None
    fallback_to: str | None = None
    fallback_reason: str | None = None
    prompt_template_key: str | None = None
    prompt_template_version: int | None = None


class ConflictStatusRecord(BaseModel):
    """Source conflict detection results."""

    detected: bool = False
    agreement_level: Literal["full", "partial", "conflicting"] = "full"
    conflict_count: int = 0
    conflicting_document_ids: list[str] = Field(default_factory=list)
    preferred_document_ids: list[str] = Field(default_factory=list)
    conflict_summary: str | None = None


class PolicyEnforcementRecord(BaseModel):
    """AI response policy outcomes (F268)."""

    applied: bool = False
    outcome: str | None = None
    violated_rules: list[str] = Field(default_factory=list)
    warning_flags: list[str] = Field(default_factory=list)
    has_disclaimer: bool = False


class SourceFreshnessRecord(BaseModel):
    """Source freshness warning summary (F297/F311)."""

    warning: bool = False
    warning_reason: str | None = None
    warning_reasons: list[str] = Field(default_factory=list)
    stale_count: int = 0
    excluded_count: int = 0
    boosted_count: int = 0
    unreviewed_count: int = 0
    deprecated_count: int = 0
    draft_count: int = 0
    all_excluded_fallback: bool = False


class EvidenceQualityRecord(BaseModel):
    """Aggregated evidence quality signals across all cited sources (F315).

    Captures table extraction confidence, document extraction health, and
    processing completeness issues that may reduce answer reliability.
    """

    table_low_confidence_count: int = 0
    extraction_warning_count: int = 0
    processing_warning_count: int = 0
    any_incomplete_documents: bool = False
    warning_reasons: list[str] = Field(default_factory=list)


class AnswerTrustMetadataResponse(BaseModel):
    """Versioned, organization-scoped answer trust metadata contract.

    schema_version is bumped when a breaking field change is introduced,
    allowing frontend consumers to branch on contract evolution without
    deployment coupling.
    """

    schema_version: Literal["1"] = "1"
    organization_id: str
    message_id: str
    not_found: bool
    citation_validation_failed: bool
    verification_failed: bool
    confidence: ConfidenceTrustRecord
    citations: list[CitationTrustRecord]
    retrieval: RetrievalDiagnosticsRecord
    query_interpretation: QueryInterpretationRecord | None = None
    grounded_verification: GroundedVerificationRecord
    model: ModelMetadataRecord
    conflict: ConflictStatusRecord
    policy: PolicyEnforcementRecord
    freshness: SourceFreshnessRecord
    evidence_quality: EvidenceQualityRecord
    generated_at: datetime


GroundedVerificationRecord.model_rebuild()
ConfidenceTrustRecord.model_rebuild()
QueryInterpretationRecord.model_rebuild()
EvidenceQualityRecord.model_rebuild()
AnswerTrustMetadataResponse.model_rebuild()

File: frontend/src/lib/api/trust_metadata.ts
/**
 * Versioned answer trust metadata contract (F307).
 *
 * Types mirror the backend AnswerTrustMetadataResponse schema exactly.
 * The schema_version field allows detecting contract changes without breaking
 * existing consumers — branch on "1" vs later versions as needed.
 *
 * No raw prompts, chain-of-thought, ACL snapshots, or internal UUIDs are
 * returned by the API; these types therefore do not include those fields.
 */

import { apiRequest } from "@/lib/api/request";

/** Normalized freshness state for UI display (F311). */
export type FreshnessState =
  | "current"
  | "stale"
  | "expired"
  | "deprecated"
  | "draft"
  | "unreviewed"
  | "unknown";

export type CitationTrustRecord = {
  document_id: string;
  chunk_id: string;
  filename?: string | null;
  page_number?: number | null;
  score?: number | null;
  similarity_score?: number | null;
  rerank_score?: number | null;
  original_rank?: number | null;
  final_rank?: number | null;
  text_snippet?: string | null;
  source_provider?: string | null;
  source_provider_label?: string | null;
  source_title?: string | null;
  source_key?: string | null;
  source_section?: string | null;
  source_deep_link?: string | null;
  source_last_synced_at?: string | null;
  source_trust_status?:
    | "trusted"
    | "stale"
    | "revoked"
    | "deleted"
    | "unknown"
    | "uploaded"
    | null;
  conflict_status?: "preferred" | "conflicting" | "neutral" | null;
  doc_trust_status?: string | null;
  doc_review_status?: string | null;
  doc_version_label?: string | null;
  doc_review_due_date?: string | null;
  doc_expiry_date?: string | null;
  doc_stale_warning: boolean;
  doc_expired_warning: boolean;
  doc_is_excluded_status: boolean;
  is_table_chunk: boolean;
  table_caption?: string | null;
  table_row_count?: number | null;
  table_col_count?: number | null;
  table_headers: string[];
  doc_ocr_quality_status?:
    | "high"
    | "medium"
    | "low"
    | "failed"
    | "not_required"
    | null;
  doc_ocr_low_confidence_warning: boolean;
  /** F311 — normalized freshness state for the trust panel badge. */
  freshness_state?: FreshnessState | null;
  /** F311 — ISO timestamp of when the source document was last modified. */
  doc_last_updated_at?: string | null;
  /** F311 — user ID of the review owner for this document. */
  doc_review_owner_id?: string | null;
  /** F311 — true when the source is pending review (needs_review status). */
  doc_unreviewed_warning: boolean;
  /** F311 — true when the source is deprecated, archived, or superseded. */
  doc_deprecated_warning: boolean;
};

/** Single explainable signal that contributed to the confidence score (F310). */
export type ConfidenceReasonRecord = {
  code: string;
  label: string;
  impact: "positive" | "negative" | "neutral";
  magnitude: number;
};

export type ConfidenceTrustRecord = {
  score: number;
  category: "low" | "medium" | "high";
  /** Calibrated trust level including quality-signal degradation (F310). */
  trust_level: "high" | "medium" | "low" | "warning" | "not_found";
  citation_support_score: number;
  citation_validation_score: number;
  citation_coverage_score: number;
  retrieval_agreement_score: number;
  top_similarity: number;
  average_similarity: number;
  top_rerank_score: number;
  raw_score: number;
  citation_validation_multiplier: number;
  not_found_penalty_multiplier: number;
  freshness_multiplier: number;
  ocr_quality_multiplier: number;
  conflict_multiplier: number;
  graph_evidence_boost: number;
  verification_support_score?: number | null;
  not_found_signal: boolean;
  no_context: boolean;
  reasons: ConfidenceReasonRecord[];
};

export type ClaimSupportRecord = {
  claim_index: number;
  claim_text: string;
  support_status:
    | "supported"
    | "partially_supported"
    | "unsupported"
    | "unverifiable";
  support_score: number;
  evidence_match_score: number;
  source_quality_score: number;
  rerank_score: number;
  chunk_coverage_score: number;
  citation_indices: number[];
};

export type RetrievalDiagnosticsRecord = {
  retrieval_candidate_count?: number;
  retrieval_count: number;
  selected_count: number;
  top_k?: number;
  search_mode?: string | null;
  source_scope_mode?: string | null;
  source_scope_label?: string | null;
  retrieval_profile_name?: string | null;
  retrieval_profile_scope?: string | null;
  retrieval_profile_source?: string | null;
  retrieval_filters?: string[];
  rerank_applied: boolean;
  rerank_provider?: string | null;
  rerank_model?: string | null;
  rerank_score_min?: number | null;
  rerank_score_max?: number | null;
  rerank_fallback_used?: boolean;
  rerank_fallback_reason?: string | null;
  request_id?: string | null;
  trace_request_id?: string | null;
  hybrid_retrieval_enabled: boolean;
  hybrid_vector_hit_count: number;
  hybrid_keyword_hit_count: number;
  query_rewriting_applied: boolean;
  query_decomposed: boolean;
  sub_query_count: number;
  parent_context_expanded_count: number;
  graph_context_used: boolean;
  graph_context_unavailable: boolean;
  graph_chunk_count: number;
  freshness_excluded_count: number;
  freshness_boosted_count: number;
};

export type QueryInterpretationRecord = {
  intent:
    | "lookup"
    | "summary"
    | "comparison"
    | "policy"
    | "troubleshooting"
    | "compliance"
    | "connector_search"
    | "graph_entity_search";
  intent_label: string;
  complexity: "simple" | "complex" | "multi_part";
  retrieval_strategy: "original" | "rewrite" | "decompose";
  rewrite_preview_enabled: boolean;
  rewritten_query_preview?: string | null;
  sub_queries: string[];
};

export type GroundedVerificationRecord = {
  applied: boolean;
  verdict?: string | null;
  score?: number | null;
  aggregate_support_score: number;
  claim_count: number;
  supported_count: number;
  partially_supported_count: number;
  unsupported_count: number;
  unverifiable_count: number;
  removed_count: number;
  reason_codes: string[];
  claims: ClaimSupportRecord[];
};

export type ModelMetadataRecord = {
  llm_model?: string | null;
  llm_provider?: string | null;
  embedding_model?: string | null;
  fallback_used: boolean;
  fallback_from?: string | null;
  fallback_to?: string | null;
  fallback_reason?: string | null;
  prompt_template_key?: string | null;
  prompt_template_version?: number | null;
};

export type ConflictStatusRecord = {
  detected: boolean;
  agreement_level: "full" | "partial" | "conflicting";
  conflict_count: number;
  conflicting_document_ids: string[];
  preferred_document_ids: string[];
  conflict_summary?: string | null;
};

export type PolicyEnforcementRecord = {
  applied: boolean;
  outcome?: string | null;
  violated_rules: string[];
  warning_flags: string[];
  has_disclaimer: boolean;
};

export type SourceFreshnessRecord = {
  warning: boolean;
  warning_reason?: string | null;
  /** F311 — structured list of specific freshness warning messages. */
  warning_reasons: string[];
  stale_count: number;
  excluded_count: number;
  boosted_count: number;
  /** F311 — number of cited sources pending review. */
  unreviewed_count: number;
  /** F311 — number of cited sources that are deprecated, archived, or superseded. */
  deprecated_count: number;
  /** F311 — true when all sources were excluded and the fallback re-included them. */
  all_excluded_fallback: boolean;
};

/** Versioned, organization-scoped answer trust metadata. schema_version "1" is current. */
export type AnswerTrustMetadataResponse = {
  schema_version: "1";
  organization_id: string;
  message_id: string;
  not_found: boolean;
  citation_validation_failed: boolean;
  verification_failed: boolean;
  confidence: ConfidenceTrustRecord;
  citations: CitationTrustRecord[];
  retrieval: RetrievalDiagnosticsRecord;
  query_interpretation?: QueryInterpretationRecord | null;
  grounded_verification: GroundedVerificationRecord;
  model: ModelMetadataRecord;
  conflict: ConflictStatusRecord;
  policy: PolicyEnforcementRecord;
  freshness: SourceFreshnessRecord;
  generated_at: string;
};

/**
 * Fetch the versioned trust metadata snapshot for a saved assistant message.
 * Returns null (404) when the message predates F307 or does not exist.
 */
export async function getAnswerTrustMetadata(
  messageId: string,
): Promise<AnswerTrustMetadataResponse> {
  return apiRequest<AnswerTrustMetadataResponse>(
    `/chat/messages/${encodeURIComponent(messageId)}/trust-metadata`,
  );
}
